### PR TITLE
Locally Connected Layer

### PR DESCRIPTION
**Pull request by joelthchao**
As #2589 proposed, locally connected layer has been used in [DeepFace](https://www.cs.toronto.edu/~ranzato/publications/taigman_cvpr14.pdf) and Alex's cifar-10 [implementation](https://code.google.com/p/cuda-convnet/source/browse/trunk/example-layers/layers-conv-local-11pct.cfg). I think it would be a plus to support this layer in Keras.
Also, I added an example which try to reproduce Alex's cifar-10 network using this layer.

### CODE REVIEW

File: examples/cifar10_alex.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+'''Train a simple deep CNN on the CIFAR10 small images dataset.
+In this example, we use LocallyConnected2D to reproduce Alex's implement
+There are two difference, one is input size (32x32 vs. 24x24), the other 
+is replacing LRN with BatchNormalization.
```
Comment (fchollet) [inline examples/cifar10_alex.py]: These two operations are significantly different though...

File: examples/cifar10_alex.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+GPU run command:
+    THEANO_FLAGS=mode=FAST_RUN,device=gpu,floatX=float32 python cifar10_cnn.py
+
+It gets down to 0.65 test logloss in 25 epochs, and down to 0.55 after 50 epochs.
```
Comment (fchollet) [inline examples/cifar10_alex.py]: What is the final accuracy of the model; are you able to reproduce the results of the original model?

File: keras/layers/local.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+                                    self.filter_length,
+                                    self.border_mode,
+                                    self.subsample_length)
+        return (input_shape[0], length, self.nb_filter)
+
+    def call(self, x, mask=None):
+        stride = self.subsample_length
+        output_length, filter_length, input_dim, nb_filter = self.W_shape
+        feature_dim = filter_length * input_dim
+        w_flatten = K.reshape(self.W, (output_length, feature_dim, nb_filter))
```
Comment (fchollet) [inline keras/layers/local.py]: Why not have `self.W` already have the proper shape? That would spare one reshape operation (which is expensive and would be run at every step).

**Comment by joelthchao** [comment]
@fchollet Thanks for your valuable comments!
> These two operations are significantly different though...

For the `BatchNormalization`, I would remove it from this example, although it gives a marginal improvement. 

>What is the final accuracy of the model; are you able to reproduce the results of the original model?

The test logloss is a little better than the `cifar10_cnn.py`. However, I can't reproduce the original model, which claims that it would have 87% test accuracy in 100 epochs. It would need to change input by cropping 24x24 patches from 32x32 image, but I prefer to keep it simple in example. BTW, do we have cropping in `ImageDataGenerator`?

val_loss
Blue: `cifar10_cnn.py`, Purple: `cifar10_alex.py`

> Why not have self.W already have the proper shape? 

Good advise, I will modify it.

**Comment by fchollet** [comment]
> However, I can't reproduce the original model, which claims that it would have 87% test accuracy in 100 epochs. 

What accuracy do you get after 100 epochs?

> BTW, do we have cropping in ImageDataGenerator?

Yes, I'm pretty sure we do.

It would be great to be able to reproduce the original results.

**Comment by joelthchao** [comment]
> What accuracy do you get after 100 epochs?

Test accuracy (center crop) will be around **81%**. The original network uses **10 views average** and **manual learning rate control** to achieve 87% accuracy (about 150 epochs).

They also define different learning rate for weight/bias and different std for weight initialization for different layers. I may need not have time to carefully reproduce it. Anyone wants to take the challenge?

Actually, we don't have cropping mechanism in the ImageDataGenerator. Therefore I use random shift in `ImageDataGenerator` and center cropping by `Lambda` Layer to achieve it.

**Comment by fchollet** [comment]
I think if we can't reproduce the 87% baseline we shouldn't include the example script as it may be misleading.

**Comment by joelthchao** [comment]
> I think if we can't reproduce the 87% baseline we shouldn't include the example script as it may be misleading.

Sure, I will remove it.

File: keras/layers/local.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+from keras import backend as K
+from keras.layers import activations, initializations, regularizers, constraints
+from keras.engine import Layer, InputSpec
+
+
+def conv_output_length(input_length, filter_size, border_mode, stride):
```
Comment (fchollet) [inline keras/layers/local.py]: This function is defined on `convolutional.py`. Why redefine it here?

File: keras/layers/local.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+                                    self.subsample_length)
+        return (input_shape[0], length, self.nb_filter)
+
+    def call(self, x, mask=None):
+        stride = self.subsample_length
+        output_length, feature_dim, nb_filter = self.W_shape
+
+        output = []
+        for i in range(output_length):
+            slice_length = slice(i*stride, i*stride + self.filter_length)
```
Comment (fchollet) [inline keras/layers/local.py]: Spaces around operators please.

File: keras/layers/local.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        output_length = input_length
+    elif border_mode == 'valid':
+        output_length = input_length - filter_size + 1
+    return (output_length + stride - 1) // stride
+
+
+class LocallyConnected1D(Layer):
+    # TODO batch_dot
+    '''LocallyConnected1D layer works almost the same as Convolution1D layer,
+    except that weights are unshared. When using this layer as the first layer
```
Comment (fchollet) [inline keras/layers/local.py]: Maybe some precision about what this means is needed, e.g. "a different weight matrix is used for each different patch of the input"...

File: keras/layers/local.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    assert border_mode in {'same', 'valid'}
+    if border_mode == 'same':
+        output_length = input_length
+    elif border_mode == 'valid':
+        output_length = input_length - filter_size + 1
+    return (output_length + stride - 1) // stride
+
+
+class LocallyConnected1D(Layer):
+    # TODO batch_dot
```
Comment (fchollet) [inline keras/layers/local.py]: I don't understand what this TODO is about?

File: keras/layers/local.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+                  'input_dim': self.input_dim,
+                  'input_length': self.input_length}
+        base_config = super(LocallyConnected1D, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+class LocallyConnected2D(Layer):
+    # TODO batch_dot
+    '''LocallyConnected2D layer works almost the same as Convolution2D layer,
+    except that weights are unshared. When using this layer as the first layer
```
Comment (fchollet) [inline keras/layers/local.py]: Same here.

File: keras/layers/local.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    def call(self, x, mask=None):
+        # TODO: border_mode
+        stride_row, stride_col = self.subsample
+        output_row, output_col, feature_dim, nb_filter = self.W_shape
+
+        if self.dim_ordering == 'th':
+            output = []
+            for i in range(output_row):
+                for j in range(output_col):
+                    slice_row = slice(i*stride_row, i*stride_row + self.nb_row)
```
Comment (fchollet) [inline keras/layers/local.py]: Spaces around operators please.

File: keras/layers/local.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+
+        if self.dim_ordering == 'th':
+            return (input_shape[0], self.nb_filter, rows, cols)
+        elif self.dim_ordering == 'tf':
+            return (input_shape[0], rows, cols, self.nb_filter)
+        else:
+            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+
+    def call(self, x, mask=None):
+        # TODO: border_mode
```
Comment (fchollet) [inline keras/layers/local.py]: How do you intend to handle border_mode?

File: keras/layers/local.py [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+            This parameter is only relevant if you don't pass
+            a `weights` argument.
+        activation: name of activation function to use
+            (see [activations](../activations.md)),
+            or alternatively, elementwise Theano function.
+            If you don't specify anything, no activation is applied
+            (ie. "linear" activation: a(x) = x).
+        weights: list of numpy arrays to set as initial weights.
+        border_mode: Only support 'valid'. Please make good use of
+            ZeroPadding2D to achieve same oupout shape.
```
Comment (fchollet) [inline keras/layers/local.py]: Typo

File: keras/layers/local.py [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+            or alternatively, Theano function to use for weights initialization.
+            This parameter is only relevant if you don't pass a `weights` argument.
+        activation: name of activation function to use
+            (see [activations](../activations.md)),
+            or alternatively, elementwise Theano function.
+            If you don't specify anything, no activation is applied
+            (ie. "linear" activation: a(x) = x).
+        weights: list of numpy arrays to set as initial weights.
+        border_mode: Only support 'valid'. Please make good use of
+            ZeroPadding1D to achieve same oupout_length.
```
Comment (fchollet) [inline keras/layers/local.py]: Typo

**Comment by joelthchao** [comment]
@fchollet Thanks for your careful review! I will push again tomorrow.

- I planed to use `K.batch_dot` to replace multiple `K.dot` on different patch of input. But I think it would potentially result huge memory usage when `batch_size` or `output_shape` is large.

- Not sure if it is necessary to support `same` mode. I prefer not haha. If needed, zero padding will be used.

- Should I modify mkdoc for this new layer?

**Comment by joelthchao** [comment]
**UPDATE**: Parallel `dot` in different position (patch) by `batch_dot` with reduction of 40% computation time on Alex's implement (heavily depends on configuration of layer).
P.S. experiment on TensorFlow.

**Comment by fchollet** [comment]
It looks like `test_locallyconnected_2d` is currently failing with both backends.

**Comment by joelthchao** [comment]
I am dealing with a theano issue, might take some time to think up a workaround.

**Comment by joelthchao** [comment]
- Currently, I can't solve the problem mention in #2742, so in the case `K._backend==theano` and `dim_ordering=='th'` I will use multiple `K.dot` instead.

- Also, update `test_local.py` to cover different `dim_ordering`. Unit test for convolutional layer might need same modification as well.

**Comment by nitish11** [comment]
I used this build and try to run a  deepnet model with keras build.

_THEANO_FLAGS=optimizer=fast_compile,device=gpu,floatX=float32 nohup sudo python model_training.py > output.out &_

I am getting the below error 

_ERROR (theano.gof.opt): Optimization failure due to: local_useless_inc_subtensor
ERROR (theano.gof.opt): node: IncSubtensor{Inc;int64, ::, ::}(Elemwise{add,no_inplace}.0, DimShuffle{0,1}.0, Constant{0})
ERROR (theano.gof.opt): TRACEBACK:
ERROR (theano.gof.opt): Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/opt.py", line 1772, in process_node
    replacements = lopt.transform(node)
  File "/usr/local/lib/python2.7/dist-packages/theano/tensor/opt.py", line 2313, in local_useless_inc_subtensor
    c = get_scalar_constant_value(node.inputs[0])
  File "/usr/local/lib/python2.7/dist-packages/theano/tensor/basic.py", line 662, in get_scalar_constant_value
    v.owner.op.perform(v.owner, const, ret)
  File "/usr/local/lib/python2.7/dist-packages/theano/tensor/elemwise.py", line 839, in perform
    super(Elemwise, self).perform(node, inputs, output_storage)
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/op.py", line 769, in perform
    "Did you used Theano flags mode=FAST_COMPILE?"
MethodNotDefined: ('perform', <class 'theano.tensor.elemwise.Elemwise'>, 'Elemwise', 'Did you used Theano flags mode=FAST_COMPILE? You can use optimizer=fast_compile instead.')_

**Comment by fchollet** [comment]
@joelthchao Hi Joel, what's the current status here? Should I review and merge?

**Comment by xyzhang16** [comment]
@jfsantos  thank you for your work. It help me a lot.

**Comment by xyzhang16** [comment]
I wrote a toy example using your layer
code:
`
x_test=np.random.random_sample((100,5))
y_test=np.random.randint(2,size=100)
model=Sequential()
model.add(LocallyConnected1D(nb_filter=1,filter_length=2,input_dim=5))
model.add(Flatten())
model.add(Dense(1))
model.compile(optimizer='sgd', loss='mse')
model.fit(x_test,y_test)
`
TypeError: an integer is required
why is it wrong?

**Comment by joelthchao** [comment]
@fchollet It should be ready. You can review it.
@xyzhang16 `input_length` is not set. The layer require a determined input shape.

**Comment by xyzhang16** [comment]
@joelthchao  sorry, still wrong. 
code is following
`x_test=np.random.random_sample((100,5))
y_test=np.random.randint(2,size=100)
model=Sequential()
model.add(local.LocallyConnected1D(nb_filter=1,filter_length=2,input_length=5,input_dim=5))
model.add(Flatten())
model.add(Dense(1))
model.compile(optimizer='sgd', loss='mse')
model.fit(x_test,y_test)`
TypeError:Bad input argument to theano function with name "C:\\Anaconda2\\lib\\site-packages\\keras\\backend\\theano_backend.py:514"  at index 0(0-based)', 'Wrong number of dimensions: expected 3, got 2 with shape (32, 5).

**Comment by xyzhang16** [comment]
@joelthchao  thank you!

File: tests/keras/layers/test_local.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+                               'W_regularizer': 'l2',
+                               'b_regularizer': 'l2',
+                               'activity_regularizer': 'activity_l2',
+                               'subsample': subsample,
+                               'dim_ordering': 'th'},
+                       input_shape=(nb_samples, stack_size, nb_row, nb_col))
+
+
+if __name__ == '__main__':
+    # pytest.main([__file__])
```
Comment (fchollet) [inline tests/keras/layers/test_local.py]: Better to use `pytest.main([__file__])` here.

**Comment by Simoncarbo** [comment]
Hi everyone, and thank you for your work!

Running test_local.py gives me the following warning:


> ...\Anaconda3\envs\python34\lib\site-packages\keras-1.0.4-py3.4.egg\keras\backend\theano_backend.py:521: UserWarning: theano.function was asked to create a function computing outputs given certain inputs, but the provided input variable at index 1 is not part of the computational graph needed to compute the outputs: keras_learning_phase.
> To make this warning into an error, you can pass the parameter on_unused_input='raise' to theano.function. To disable it completely, use on_unused_input='ignore'.
>   **kwargs)


And trying a simple model:

```
    model = Sequential()
       
    model.add(layers.ZeroPadding2D(padding=(1, 1),input_shape = (1,nb_rows,nb_columns))) 
    model.add(layers.LocallyConnected2D(2,3,3,activation = 'relu',border_mode = 'valid'))
    
    model.add(layers.Flatten(input_shape = (1,nb_rows,nb_columns)))
    model.add(layers.Dense(100, activation = 'relu')) #,W_regularizer=l2(0.01)
    model.add(layers.BatchNormalization(mode = 0,axis = 1))
    model.add(layers.Dropout(0.2))
    model.add(layers.Dense(80, activation = 'relu'))
    model.add(layers.BatchNormalization(mode = 0,axis = 1))
    model.add(layers.Dropout(0.1))
    model.add(layers.Dense(40, activation = 'relu'))
    model.add(layers.BatchNormalization(mode = 0,axis = 1))
    model.add(layers.Dropout(0.1))
    model.add(layers.Dense(10, activation = 'relu'))
    model.add(layers.BatchNormalization(mode = 0,axis = 1))
    model.add(layers.Dense(1, activation = 'sigmoid'))
```

gives me the following (truncated) error:

```
File "...\Anaconda3\envs\python34\lib\site-packages\theano\sandbox\cuda\nvcc_compiler.py", line 403, in compile_str
    'for cmd', ' '.join(cmd))

Exception: ('The following error happened while compiling the node', GpuJoin(TensorConstant{0}, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, Rebroadcast{1}.0, ...
... ), '\n', 'nvcc return status', 2, 'for cmd', 'nvcc -shared -O3 -arch=sm_50 -Xlinker /DEBUG -D HAVE_ROUND -m64 -Xcompiler -DCUDA_NDARRAY_CUH=m18715462c72ed6afcd7ca5d52813ce90,-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION,/Zi,/MD -IC:\\Users\\sicarbonnell\\AppData\\Local\\Theano\\compiledir_Windows-7-6.1.7601-SP1-Intel64_Family_6_Model_61_Stepping_4_GenuineIntel-3.4.4-64\\cuda_ndarray -IC:\\Users\\sicarbonnell\\AppData\\Local\\Continuum\\Anaconda3\\envs\\python34\\lib\\site-packages\\numpy\\core\\include -IC:\\Users\\sicarbonnell\\AppData\\Local\\Continuum\\Anaconda3\\envs\\python34\\include -IC:\\Users\\sicarbonnell\\AppData\\Local\\Continuum\\Anaconda3\\envs\\python34\\lib\\site-packages\\theano\\gof -IC:\\Users\\sicarbonnell\\AppData\\Local\\Continuum\\Anaconda3\\envs\\python34\\lib\\site-packages\\theano\\sandbox\\cuda -o C:\\Users\\sicarbonnell\\AppData\\Local\\Theano\\compiledir_Windows-7-6.1.7601-SP1-Intel64_Family_6_Model_61_Stepping_4_GenuineIntel-3.4.4-64\\tmp9f9odf_w\\mac87adaa1b92aafdf1d5bd49bc08bc35.pyd mod.cu -LC:\\Users\\sicarbonnell\\AppData\\Local\\Theano\\compiledir_Windows-7-6.1.7601-SP1-Intel64_Family_6_Model_61_Stepping_4_GenuineIntel-3.4.4-64\\cuda_ndarray -LC:\\Users\\sicarbonnell\\AppData\\Local\\Continuum\\Anaconda3\\envs\\python34\\libs -LC:\\Users\\sicarbonnell\\AppData\\Local\\Continuum\\Anaconda3\\envs\\python34 -lcudart -lcublas -lcuda_ndarray -lpython34', '[GpuJoin(TensorConstant{0}, <CudaNdarrayType(float32, (True, False, False))>, <CudaNdarrayType(float32, (True, False, False))>, <CudaNdarrayType(float32, (True, False, False))>, <CudaNdarrayType(float32, (True, False, False))>, <CudaNdarrayType(float32, (True, False, False))>, <CudaNdarrayType(float32, (True, False, False))>, ...)
```

Other models without locally connected layers are working fine. Can you help me on this?

More info: 
THEANO_FLAGS = floatX=float32,device=gpu,lib.cnmem=0.8,optimizer=fast_run
Using gpu device 0: GeForce GTX 770 (CNMeM is enabled with initial size: 80.0% of memory, cuDNN 5005)

File: keras/layers/local.py [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+    '''
+    def __init__(self, nb_filter, filter_length,
+                 init='uniform', activation='linear', weights=None,
+                 border_mode='valid', subsample_length=1,
+                 W_regularizer=None, b_regularizer=None, activity_regularizer=None,
+                 W_constraint=None, b_constraint=None,
+                 bias=True, input_dim=None, input_length=None, **kwargs):
+
+        if border_mode not in {'valid'}:
+            raise Exception('Invalid border mode for Convolution1D:', border_mode)
```
Comment (fchollet) [inline keras/layers/local.py]: Maybe better for the check to be `!= 'valid'` and the error message to say "only border_mode "valid" is supported", unless you have plans to support more border modes.

File: keras/layers/local.py [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+        self.nb_filter = nb_filter
+        self.filter_length = filter_length
+        self.init = initializations.get(init, dim_ordering='th')
+        self.activation = activations.get(activation)
+        assert border_mode in {'valid'}, 'border_mode must be in {valid, same}'
```
Comment (fchollet) [inline keras/layers/local.py]: This check should be remove (and the error message doesn't match the check).

File: keras/layers/local.py [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+    def call(self, x, mask=None):
+        stride = self.subsample_length
+        output_length, feature_dim, nb_filter = self.W_shape
+
+        xs = []
+        for i in range(output_length):
+            slice_length = slice(i * stride, i * stride + self.filter_length)
+            xs.append(K.reshape(x[:, slice_length, :], (1, -1, feature_dim)))
+        x_aggregate = K.concatenate(xs, axis=0)
+        output = K.batch_dot(x_aggregate, self.W) # (output_length, batch_size, nb_filter)
```
Comment (fchollet) [inline keras/layers/local.py]: In-line comments should be separated by two spaces (`  #`)

File: keras/layers/local.py [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+                 bias=True, **kwargs):
+
+        if border_mode not in {'valid'}:
+            raise Exception('Invalid border mode for Convolution2D:', border_mode)
+        self.nb_filter = nb_filter
+        self.nb_row = nb_row
+        self.nb_col = nb_col
+        self.init = initializations.get(init, dim_ordering=dim_ordering)
+        self.activation = activations.get(activation)
+        assert border_mode in {'valid'}, 'border_mode must be in {valid}'
```
Comment (fchollet) [inline keras/layers/local.py]: Should be removed

File: keras/layers/local.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+    '''
+    def __init__(self, nb_filter, nb_row, nb_col,
+                 init='glorot_uniform', activation='linear', weights=None,
+                 border_mode='valid', subsample=(1, 1), dim_ordering='th',
+                 W_regularizer=None, b_regularizer=None, activity_regularizer=None,
+                 W_constraint=None, b_constraint=None,
+                 bias=True, **kwargs):
+
+        if border_mode not in {'valid'}:
+            raise Exception('Invalid border mode for Convolution2D:', border_mode)
```
Comment (fchollet) [inline keras/layers/local.py]: Same comment as above

File: keras/layers/local.py [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+    # Output shape
+        4D tensor with shape:
+        `(samples, nb_filter, new_rows, new_cols)` if dim_ordering='th'
+        or 4D tensor with shape:
+        `(samples, new_rows, new_cols, nb_filter)` if dim_ordering='tf'.
+        `rows` and `cols` values might have changed due to padding.
+    '''
+    def __init__(self, nb_filter, nb_row, nb_col,
+                 init='glorot_uniform', activation='linear', weights=None,
+                 border_mode='valid', subsample=(1, 1), dim_ordering='th',
```
Comment (fchollet) [inline keras/layers/local.py]: `dim_ordering` should default to `K.image_dim_ordering()`.

**Comment by fchollet** [comment]
LGTM overall. Please address the comments I left, otherwise looking good!

**Comment by fchollet** [comment]
I have applied the last changes and merged this PR: https://github.com/fchollet/keras/commit/835a02c037384b4ccf0e5b2425c6dd7469bf702d

Thanks for the great new layers.

**Comment by joelthchao** [comment]
@fchollet  Sorry I am too busy to deal with it, thanks for your great help!

**Pull request closed without merging**